flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -241,28 +241,131 @@ Fortran::lower::FirOpBuilder::createShape(mlir::Location loc,
+            .readLengthFromBox(x.getAddr());
+      },
+      [&](const auto &) -> mlir::Value {
+        fir::emitFatalError(
+            loc, "Character length inquiry on a non-character entity");
+      });
+}
+
+mlir::Value Fortran::lower::FirOpBuilder::readExtent(
+    mlir::Location loc, const fir::ExtendedValue &box, unsigned dim) {
+  assert(box.rank() > dim);
+  return box.match(
+      [&](const fir::ArrayBoxValue &x) -> mlir::Value {
+        return x.getExtents()[dim];
+      },
+      [&](const fir::CharArrayBoxValue &x) -> mlir::Value {
+        return x.getExtents()[dim];
+      },
+      [&](const fir::BoxValue &x) -> mlir::Value {
+        return x.getExtents()[dim];
+      },
+      [&](const fir::IrBoxValue &x) -> mlir::Value {
+        if (!x.getExplicitExtents().empty())
+          return x.getExplicitExtents()[dim];
+        auto idxTy = getIndexType();
+        auto dimVal = createIntegerConstant(loc, idxTy, dim);
+        return create<fir::BoxDimsOp>(loc, idxTy, idxTy, idxTy, x.getAddr(),
+                                      dimVal)
+            .getResult(1);
+      },
+      [&](const auto &) -> mlir::Value {
+        fir::emitFatalError(loc, "extent inquiry on scalar");
+      });
+}
+
+mlir::Value Fortran::lower::FirOpBuilder::readLowerBound(
+    mlir::Location loc, const fir::ExtendedValue &box, unsigned dim,
+    mlir::Value defaultValue) {
+  assert(box.rank() > dim);
+  auto lb = box.match(
+      [&](const fir::ArrayBoxValue &x) -> mlir::Value {
+        return x.getLBounds().empty() ? mlir::Value{} : x.getLBounds()[dim];
+      },
+      [&](const fir::CharArrayBoxValue &x) -> mlir::Value {
+        return x.getLBounds().empty() ? mlir::Value{} : x.getLBounds()[dim];
+      },
+      [&](const fir::BoxValue &x) -> mlir::Value {
+        return x.getLBounds().empty() ? mlir::Value{} : x.getLBounds()[dim];
+      },
+      [&](const fir::IrBoxValue &x) -> mlir::Value {
+        return x.getLBounds().empty() ? mlir::Value{} : x.getLBounds()[dim];
+      },
+      [&](const auto &) -> mlir::Value {
+        fir::emitFatalError(loc, "lower bound inquiry on scalar");
+      });
+  if (lb)
+    return lb;
+  if (defaultValue)
+    return defaultValue;
+  return createIntegerConstant(loc, getIndexType(), 1);
+}
+
+void Fortran::lower::FirOpBuilder::readExtents(
+    mlir::Location loc, const fir::IrBoxValue &box,
+    llvm::SmallVectorImpl<mlir::Value> &result) {
+  assert(result.empty());
+  auto explicitExtents = box.getExplicitExtents();
+  if (!explicitExtents.empty()) {
+    result.append(explicitExtents.begin(), explicitExtents.end());
+    return;
+  }
+  auto rank = box.rank();
+  auto idxTy = getIndexType();
+  for (decltype(rank) dim = 0; dim < rank; ++dim) {
+    auto dimVal = createIntegerConstant(loc, idxTy, dim);
+    auto dimInfo =
+        create<fir::BoxDimsOp>(loc, idxTy, idxTy, idxTy, box.getAddr(), dimVal);
+    result.emplace_back(dimInfo.getResult(1));
+  }
+}
+
```

Comment:
See earlier comment. These may be a bit more "macro" is granularity than one might expect in a Builder.

Reply:
Yes. I moved them out of the class.